From 0555956825b3788fd30f61a368f83d657b77e6aa Mon Sep 17 00:00:00 2001
From: Melissa Wen <mwen@igalia.com>
Date: Mon, 2 Feb 2026 22:26:33 -0100
Subject: [PATCH 032/188] drm/amd/display: rework segmentation for linearizing
 LUT
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

When linearizing (degamma), input is already in a near-linear space, so
the best distribution is a uniform one. Fix region/segment distribution
for PQ/sRGB pre-defined degamma TF by halving the #pts as you go down,
so 128 pts(segments) in the top region, then 64 in the next smaller, 32
etc. This way the spacing in [0.5,1] region is (1-0.5)/128, in
[0.25,0.5] region it's (0.5-0.25)/64 == 0.5/128, and so on. This fix
banding when userspace set blend/post-3DLUT curve to PQ predefined TF.

Suggested-by: Krunoslav Kovac <Krunoslav.Kovac@amd.com>
Signed-off-by: Melissa Wen <mwen@igalia.com>
(cherry picked from commit 6074a410d71fdd05f42897e8225254596384d9d2)
Signed-off-by: Nícolas F. R. A. Prado <nfraprado@collabora.com>
---
 .../amd/display/dc/dcn30/dcn30_cm_common.c    | 35 +++++++++++++------
 1 file changed, 25 insertions(+), 10 deletions(-)

diff --git a/drivers/gpu/drm/amd/display/dc/dcn30/dcn30_cm_common.c b/drivers/gpu/drm/amd/display/dc/dcn30/dcn30_cm_common.c
index 85b441f0c..da07c6e05 100644
--- a/drivers/gpu/drm/amd/display/dc/dcn30/dcn30_cm_common.c
+++ b/drivers/gpu/drm/amd/display/dc/dcn30/dcn30_cm_common.c
@@ -303,9 +303,6 @@ bool cm3_helper_translate_curve_to_hw_format(struct dc_context *ctx,
 	return true;
 }
 
-#define NUM_DEGAMMA_REGIONS    12
-
-
 bool cm3_helper_translate_curve_to_degamma_hw_format(
 				const struct dc_transfer_func *output_tf,
 				struct pwl_params *lut_params)
@@ -329,17 +326,33 @@ bool cm3_helper_translate_curve_to_degamma_hw_format(
 	memset(lut_params, 0, sizeof(struct pwl_params));
 	memset(seg_distr, 0, sizeof(seg_distr));
 
-	region_start = -NUM_DEGAMMA_REGIONS;
-	region_end   = 0;
+	if (output_tf->tf == TRANSFER_FUNCTION_PQ || TRANSFER_FUNCTION_SRGB) {
+		/* 9 segments
+		 * segments are from 2^-8 to 0
+		 */
+		const uint8_t SEG_COUNT = 9;
+		seg_distr[0] = 0; // Since we only have one point in last region
+		for (k = 1; k < SEG_COUNT; k++)
+			seg_distr[k] = k - 1; // Depends upon the regions' points 2^(n-1); seg_distr = n
 
+		region_start = -SEG_COUNT;
+		region_end = 0;
+	} else {
+		/* 12 segments
+		 * segments are from 2^-11 to 2^0
+		 * There are less than 256 points, for optimization
+		 */
+		const uint8_t SEG_COUNT = 12;
+
+		for (i = 0; i < SEG_COUNT; i++)
+			seg_distr[i] = 4;
+
+		region_start = -SEG_COUNT;
+		region_end = 0;
+	}
 
 	for (i = region_end - region_start; i < MAX_REGIONS_NUMBER ; i++)
 		seg_distr[i] = -1;
-	/* 12 segments
-	 * segments are from 2^-12 to 0
-	 */
-	for (i = 0; i < NUM_DEGAMMA_REGIONS ; i++)
-		seg_distr[i] = 4;
 
 	for (k = 0; k < MAX_REGIONS_NUMBER; k++) {
 		if (seg_distr[k] != -1)
@@ -400,6 +413,7 @@ bool cm3_helper_translate_curve_to_degamma_hw_format(
 					     dc_fixpt_from_int(region_start));
 	corner_points[0].green.x = corner_points[0].red.x;
 	corner_points[0].blue.x = corner_points[0].red.x;
+
 	corner_points[1].red.x = dc_fixpt_pow(dc_fixpt_from_int(2),
 					     dc_fixpt_from_int(region_end));
 	corner_points[1].green.x = corner_points[1].red.x;
@@ -473,6 +487,7 @@ bool cm3_helper_translate_curve_to_degamma_hw_format(
 		++rgb;
 		++i;
 	}
+
 	cm3_helper_convert_to_custom_float(rgb_resulted,
 						lut_params->corner_points,
 						hw_points, false);
-- 
2.55.0

